knotx-knot/knotx-knot-handlebars/src/main/java/io/knotx/knot/templating/impl/HandlebarsKnotProxyImpl.java

```diff
@@ -15,39 +15,63 @@
  */
 package io.knotx.knot.templating.impl;
 
+import org.apache.commons.lang3.StringUtils;
+
+import com.github.jknack.handlebars.Context;
 import com.github.jknack.handlebars.Handlebars;
+import com.github.jknack.handlebars.Template;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.Sets;
+
 import io.knotx.dataobjects.ClientResponse;
+import io.knotx.dataobjects.Fragment;
 import io.knotx.dataobjects.KnotContext;
+import io.knotx.fragments.FragmentContentExtractor;
 import io.knotx.knot.AbstractKnotProxy;
 import io.knotx.knot.templating.HandlebarsKnotConfiguration;
 import io.knotx.knot.templating.handlebars.CustomHandlebarsHelper;
+import io.knotx.knot.templating.handlebars.JsonObjectValueResolver;
 import io.knotx.knot.templating.helpers.DefaultHandlebarsHelpers;
 import io.netty.handler.codec.http.HttpResponseStatus;
 import io.reactivex.Single;
+import io.vertx.core.json.JsonObject;
 import io.vertx.core.logging.Logger;
 import io.vertx.core.logging.LoggerFactory;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
 import java.util.Optional;
 import java.util.ServiceLoader;
 import java.util.Set;
-import org.apache.commons.lang3.StringUtils;
+import java.util.concurrent.ExecutionException;
 
 public class HandlebarsKnotProxyImpl extends AbstractKnotProxy {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(HandlebarsKnotProxyImpl.class);
 
-  private static final String START_WEBSERVICE_CALL_DEBUG_MARKER = "<!-- start compiled snippet -->";
-
-  private static final String END_WEBSERVICE_CALL_DEBUG_MARKER = "<!-- end compiled snippet -->";
-
   private static final String SUPPORTED_FRAGMENT_KNOT = "handlebars";
 
+  private static final int DEBUG_MAX_FRAGMENT_CONTENT_LOG_LENGTH = 256;
+
   private Handlebars handlebars;
 
-  private HandlebarsKnotConfiguration configuration;
+  private Cache<String, Template> cache;
+
+  private MessageDigest digest;
 
   public HandlebarsKnotProxyImpl(HandlebarsKnotConfiguration configuration) {
-    this.configuration = configuration;
     this.handlebars = createHandlebars();
+    this.cache = CacheBuilder.newBuilder()
+        .maximumSize(configuration.getCacheSize())
+        .build();
+    try {
+      this.digest = MessageDigest.getInstance(configuration.getCacheKeyAlgorithm());
+    } catch (NoSuchAlgorithmException e) {
+      LOGGER.error("Could not initialize fragment hashing algorithm!", e);
+      throw new IllegalStateException(e);
```

Comment:
It doesn't look like illegal state.
I think more appropriate here would be `IllegalArgumentException`.

Reply:
fixed